enterprise/server/auth/auth.go

```diff
@@ -359,17 +372,32 @@ func createAuthenticatorsFromConfig(ctx context.Context, authConfigs []config.Oa
 			ClientID:     authConfig.ClientID,
 			ClientSecret: authConfig.ClientSecret,
 			RedirectURL:  authURL.String(),
-			Endpoint:     provider.Endpoint(),
 			// "openid" is a required scope for OpenID Connect flows.
 			Scopes: []string{oidc.ScopeOpenID, "profile", "email"},
 		}
-		authenticators = append(authenticators, &oidcAuthenticator{
+		var once sync.Once
+		authenticator := &oidcAuthenticator{
 			slug:         authConfig.Slug,
 			issuer:       authConfig.IssuerURL,
 			oauth2Config: oauth2Config,
 			oidcConfig:   oidcConfig,
-			provider:     provider,
-		})
+		}
+		// initialize on-demand, since our mock oauth provider won't be reachable until the server starts
```

Comment:
How about having an internal bool to prevent it from running more than once?

```
AddHealthCheck(..., func(ctx ...) error {
  if initialized {
    return nil
  }
  if err := auth.Initialize(); err != nil {
    return err
  }
  initialized = true
  return nil
})
```

(a util func in `healthcheck.go` like `AddStartupCheck(healthChecker, func)` could do this automatically)

Reply:
Chatted offline a bit -- the lazy-init approach is fine (it's simpler than a health-check), but one problem is that if the oidc provider has a blip, the server gets stuck in a broken state while reporting itself as healthy, because we cache the error.

One way around this:
* Do the lazy init only for the mock (selfauth?) case -- otherwise initialize immediately in the constructor. This way, the server will do what it does today if the OIDC provider fails to be configured (it crashes on startup and K8S will CrashLoopBackoff, restarting the server until it succeeds)

* If we are worried about errors in the selfauth case as well (i.e. a temporary internal error), then one way around it is to implement the lazy-init so that it can run as many times as is needed until there is no error (can maybe implement this with a mutex, rather than sync.Once)

Reply:
Seems OK to me since it fixes the problem that the server can get stuck in a broken state.

One small downside is that if people misconfigure their auth provider, they no longer get the behavior that the server crashes on startup and they see the error printed immediately. I'm not sure how common that problem is in practice though, so I'd be fine to keep as-is :shrug: